### PR TITLE
feat(components): exponential backoff for resilient components

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Checklist prior to review

Introduce exponential backoff when resilient component fails.
Components that are using resilient component:
* KDS
* OTEL
* Postgres events
* Zone watcher
* Insights resync

In all cases, it makes sense to have exponential backoff because those components relies on external systems - API/DB etc.

I tested it manually.

- [X] [Link to relevant issue][1] as well as docs and UI issues --
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Review by slonka — COMMENTED** [review]
A bit scary to do a change like this that touches so many parts of the system without any tests changing 🤔 Are we sure there aren't any resilient components in parent project that could be impacted by this?

**Comment by jakubdyszkiewicz** [comment]
There are, but by principle those are also I/O, so they can fail the same way - an external system is not available.

File: pkg/core/runtime/component/resilient.go [hunk 1]
```diff
@@ -5,10 +5,12 @@ import (
 
 	"github.com/go-logr/logr"
 	"github.com/pkg/errors"
+	"github.com/sethvargo/go-retry"
 )
 
 const (
-	backoffTime = 5 * time.Second
+	backoffBaseTime = 5 * time.Second
+	backoffMaxTime  = 1 * time.Minute
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/core/runtime/component/resilient.go:13]: We could, but why is KDS special?
If we go with 5mins for KDS, why 5mins is not ok, but 1min is ok for insight resync for example?

[the remark this answers is not in the document]
Comment (slonka) [reply pkg/core/runtime/component/resilient.go:13]: I think that we're the most exposed to KDS errors via parent project hosted platform, so it's just for convenience. [Redis's max backoff](https://cloud.google.com/memorystore/docs/redis/exponential-backoff) is 64 seconds so 1m is close enough.

File: pkg/core/runtime/component/resilient.go [hunk 2]
```diff
@@ -51,7 +54,8 @@ func (r *resilientComponent) Start(stop <-chan struct{}) error {
 				r.log.WithValues("generationID", generationID).Error(err, "component terminated with an error")
 			}
 		}
-		<-time.After(backoffTime)
+		dur, _ := backoff.Next()
```
Comment (lahabana) [inline pkg/core/runtime/component/resilient.go:57]: Do we need to reset this backoff if we've worked fine for a while? I'd say if the component stays up for more than the max backoff reseting the backoff sounds good.

File: pkg/core/runtime/component/resilient.go [hunk 3]
```diff
@@ -25,6 +27,7 @@ func NewResilientComponent(log logr.Logger, component Component) Component {
 
 func (r *resilientComponent) Start(stop <-chan struct{}) error {
 	r.log.Info("starting resilient component ...")
+	backoff := retry.WithCappedDuration(backoffMaxTime, retry.NewExponential(backoffBaseTime))
```
Comment (lahabana) [inline pkg/core/runtime/component/resilient.go:30]:
```suggestion
	backoff := retry.WithJitter(retry.WithCappedDuration(backoffMaxTime, retry.NewExponential(backoffBaseTime)), backoffBaseTime)
```

Can't hurt no?

File: pkg/core/runtime/component/resilient.go [hunk 4]
```diff
@@ -42,23 +42,38 @@ func (r *resilientComponent) Start(stop <-chan struct{}) error {
...
+			case err := <-errCh:
+				if err != nil {
+					r.log.WithValues("generationID", generationID).Error(err, "component terminated with an error")
+				}
+				receivedErr = true
+			case <-time.After(backoffMaxTime):
+				// reset backoff so in a case of events
+				// 1) Component is unhealthy until max backoff is reached
+				// 2) Component is healthy for at least backoff max time
+				// 3) Component is unhealthy
```
Comment (slonka) [inline pkg/core/runtime/component/resilient.go:63]: I don't understand 1 and 3 - IMO we should only reset backoff when the component is healthy again. Also isn't 3 just superset of 1 so we don't need 1?

Comment (jakubdyszkiewicz) [reply pkg/core/runtime/component/resilient.go:63]: it's steps. First step then second step then third

File: pkg/core/runtime/component/resilient.go [hunk 5]
```diff
@@ -25,6 +27,7 @@ func NewResilientComponent(log logr.Logger, component Component) Component {
 
 func (r *resilientComponent) Start(stop <-chan struct{}) error {
 	r.log.Info("starting resilient component ...")
+	backoff := newBackoff()
 	for generationID := uint64(1); ; generationID++ {
 		errCh := make(chan error, 1)
```
Comment (lahabana) [inline pkg/core/runtime/component/resilient.go:32]: lastStart := time.Now()

File: pkg/core/runtime/component/resilient.go [hunk 6]
```diff
@@ -39,22 +42,38 @@ func (r *resilientComponent) Start(stop <-chan struct{}) error {
...
-
 			errCh <- r.component.Start(stop)
 		}(errCh)
-		select {
-		case <-stop:
-			r.log.Info("done")
-			return nil
-		case err := <-errCh:
-			if err != nil {
-				r.log.WithValues("generationID", generationID).Error(err, "component terminated with an error")
```
Comment (lahabana) [inline pkg/core/runtime/component/resilient.go:51]: if lastStart.Sub(time.Now)) > backoffMaxTime {
   backoff = newBackoff()
}

**Pull request merged**